trino/trino.go

```diff
@@ -1253,13 +1264,239 @@ type queryResponse struct {
+	if rowsCountAssertion, ok := metadata["rowsCount"].(json.Number); ok {
+		if val, err := rowsCountAssertion.Int64(); err == nil {
+			rowsCount = val
+		}
+	}
+
+	return spoolingMetadata{
+		RowOffset:        rowOffset,
+		RowsCount:        rowsCount,
+		SegmentSize:      segmentSize,
+		UncompressedSize: uncompressedSize,
+	}, nil
+}
+
+func (sp *spoolingProtocol) fetch() ([]queryData, error) {
+	var queryData []queryData
+	for _, segment := range sp.segments {
+		segment, ok := segment.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("segment is invalid")
+		}
+		typedMetadata, ok := segment["metadata"].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("metadata missing or invalid in segment")
+		}
+		metadata, err := parseSpoolingMetadata(typedMetadata)
+		if err != nil {
+			return nil, err
+		}
+		switch segment["type"] {
+		case "inline":
+			decodedBytes, err := base64.StdEncoding.DecodeString(segment["data"].(string))
+
+			if err != nil {
+				return nil, fmt.Errorf("error decoding base64: %v", err)
+			}
+
+			decodedData, err := decode(decodedBytes, sp.encoding, metadata)
+			if err != nil {
+				return nil, err
+			}
+
+			queryData = append(queryData, decodedData...)
+		case "spooled":
+			uri, _ := segment["uri"].(string)
+			ackUri, _ := segment["ackUri"].(string)
+			headers, _ := segment["headers"].(map[string]interface{})
+
+			data, err := sp.fetchSegment(uri, ackUri, headers)
+			if err != nil {
+				return nil, err
+			}
+
+			decodedData, err := decode(data, sp.encoding, metadata)
+			if err != nil {
+				return nil, err
+			}
+
+			queryData = append(queryData, decodedData...)
+		}
+
+	}
+
+	return queryData, nil
+}
+
+func (sp *spoolingProtocol) fetchSegment(uri, ackUri string, headers map[string]interface{}) ([]byte, error) {
+	req, err := http.NewRequestWithContext(sp.ctx, "GET", uri, nil)
+	if err != nil {
+		return nil, fmt.Errorf("trino: %w", err)
+	}
+
+	for k, v := range headers {
+		headerSlice, ok := v.([]interface{})
+
+		if ok {
+			if len(headerSlice) > 1 {
+				return nil, fmt.Errorf("trino: multiple values for header %s", k)
+			}
+			req.Header.Add(k, headerSlice[0].(string))
+			continue
+		}
+
+		return nil, fmt.Errorf("trino: unsupported header type %T", v)
+	}
+
+	resp, err := sp.httpClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("trino: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("trino: unexpected status code %d", resp.StatusCode)
+	}
+
+	data, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response body: %v", err)
+	}
+
+	//acknowledge the segment read
+	go func() {
+		ackReq, err := http.NewRequestWithContext(sp.ctx, "GET", ackUri, nil)
+		if err != nil {
+			return
+		}
+
+		for k, values := range req.Header {
+			for _, v := range values {
+				ackReq.Header.Add(k, v)
+			}
+		}
+
+		resp, err := sp.httpClient.Do(ackReq)
+		if err != nil {
```

Comment:
We should handle these errors somehow. Even if they're not actionable (acks are optional), there should be a way for the users to be aware that some acks are failing. We could send them to some buffered channel that would collect them and either return to the user or discard when the whole statement is being closed. Only ignore them when the buffered channel is full.

This could be done in the same way how the driver allows monitoring the query progress via a callback.

Since it sounds like a significant amount of work, we can do it as a follow up.